specs/merge/fork-choice.md

```diff
@@ -0,0 +1,113 @@
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+  - [Helpers](#helpers)
+    - [`PowBlock`](#powblock)
+    - [`get_pow_block`](#get_pow_block)
+    - [`is_valid_transition_block`](#is_valid_transition_block)
+  - [Updated fork-choice handlers](#updated-fork-choice-handlers)
+    - [`on_block`](#on_block)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This is the modification of the fork choice according to the executable beacon chain proposal.
+
+*Note*: It introduces the process of transition from the last PoW block to the first PoS block.
+
+### Helpers
+
+#### `PowBlock`
+
+```python
+class PowBlock(Container):
+    is_processed: boolean
+    is_valid: boolean
+    total_difficulty: uint256
+```
+
+#### `get_pow_block`
+
+Let `get_pow_block(hash: Bytes32) -> PowBlock` be the function that given the hash of the PoW block returns its data.
+
+*Note*: The `eth_getBlockByHash` JSON-RPC method does not distinguish invalid blocks from blocks that hasn't been processed yet. Either extending of existing method or implementing a new one is required.
+
+#### `is_valid_transition_block`
+
+Used by fork-choice handler, `on_block`
+
+```python
+def is_valid_transition_block(block: PowBlock) -> boolean:
+    is_total_difficulty_reached = block.total_difficulty >= TRANSITION_TOTAL_DIFFICULTY
+    return block.is_processed and block.is_valid and is_total_difficulty_reached
+```
+
+### Updated fork-choice handlers
+
+#### `on_block`
+
+*Note*: The only modification is the addition of the verification of transition block conditions.
+
+```python
+def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
+    block = signed_block.message
+    # Parent block must be known
+    assert block.parent_root in store.block_states
+    # Make a copy of the state to avoid mutability issues
+    pre_state = copy(store.block_states[block.parent_root])
+    # Blocks cannot be in the future. If they are, their consideration must be delayed until the are in the past.
+    assert get_current_slot(store) >= block.slot
+
+    # Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)
+    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
+    assert block.slot > finalized_slot
+    # Check block is a descendant of the finalized block at the checkpoint finalized slot
+    assert get_ancestor(store, block.parent_root, finalized_slot) == store.finalized_checkpoint.root
+    
+    # [Added in Merge] Consider delaying the beacon block processing until PoW block is accepted by the application node
+    if is_transition_block(pre_state, block.body):
+        pow_block = get_pow_block(block.body.application_payload.block_hash)
+        assert is_valid_transition_block(pow_block)
```

Comment:
Do you mean something like that?

```python
# Consider delaying the beacon block processing until PoW block is processed by the application layer
assert pow_block.is_processed
assert is_valid_transition_block(pow_block)
```

Reply:
Yeah, that looks good. 

A note on the comment. "consider delaying" and "delay consideration" has a surprisingly different meaning. "Consider delaying" is optional whereas "delay consideration" is more of a directive and less optional.

I would write the comment:
> Delay consideration of `block` until PoW block is processed by the PoW node

and I would put it *inside* of the `if is_transition_block` rather than on the outside
*